### DISCUSSION TITLE
GitHub integration

### DISCUSSION THREAD

**Discussion by RokCoder** [ScratchAddons/ScratchAddons#1995] [Ideas]
Integration into a (usable) GitHub format would be absolutely fantastic. And, having put some thought into this, it isn't as far-fetched an idea as it seems.

As a brief outline, a commit/push button would -

1) Separate sprites (and stage) into different source files
2) At a minimum, prettify the JSON or (ideally) parse and present it in a clearly readable non-JSON style
3) Pull extraneous information (positional, watch states, etc) into a separate dump file that doesn't really need to be readable

And an update/pull button would simply perform the inverse.

**Comment by FunctionalMetatable** [comment]
got ninja'd by 10 minutes #1996

**Comment by RokCoder** [comment]
I'd actually be happy if the commit process committed the full sb3 alongside the parsed source files and the pull process pulled the sb3 and dumped the source files. That would simplify things, ensure the sb3 couldn't be corrupted during the process and give you at a glance comparisons between versions. It's not going to give you all the bells and whistles that you can do with regular source code but that's not really what this would be about. A useful history, being able to see changes that may have caused issues and the ability to tag and use specific versions would be great.

**Reply by kriblo** [reply]
Sounds like a good starting point. Even if you can't (easily) see exactly what changed, the ability to tag versions and to see what sprites are new or changed, might be very useful (particularly in a project with many sprites).

**Comment by mxmou** [comment]
Dragging a script around in the editor changes the execution order of scripts. Ideally a system like this would sort the scripts based on something else (position in the workspace?) and store the order separately, otherwise even minor changes would have large diffs.

**Reply by TheColaber** [reply]
> Dragging a script around in the editor changes the execution order of scripts

Is that really true? According to this: https://www.youtube.com/watch?v=t9oAZs9i7jI&t=186s

**Reply by RokCoder** [reply]
> Dragging a script around in the editor changes the execution order of scripts. Ideally a system like this would sort the scripts based on something else (position in the workspace?) and store the order separately, otherwise even minor changes would have large diffs.

Alphabetical sorting would do for me. It might not be happy if you rename a custom block or similar but Beyond Compare (and other such tools) are going to give good enough results even in those cases.

**Reply by apple502j** [reply]
That was a behavior in Scratch 2.x; Scratch 3.x script execution order relies on sprite layer and when the block was added (= sorting of blocks object), coordinates don't matter

**Comment by towerofnix** [comment]
Building on https://github.com/ScratchAddons/ScratchAddons/discussions/1995#discussioncomment-542616: For implementing a representation format which is more diff-able (and readable!), [sb-edit](https://github.com/PullJosh/sb-edit/) (the backend for [Leopard](https://leopardjs.now.sh/)) would probably be a useful foundation! SB3, Leopard, and (basic) scratchblocks output formats have all been added, and it works 100% for SB3 import too, so if one wanted to add the ability to load from the generated format (e.g. to theoretically support pull requests and the like), the basic library is there to make it doable. :)

**Reply by TheColaber** [reply]
I agree, sb-edit would give a more readable interface rather than a sb3 file. In that case, would there be 2 different folders where one is the editable interface, and the other as the uploadable file?

**Reply by TheColaber** [reply]
As far I know... this is an sb3 file... 
![image](https://user-images.githubusercontent.com/72760579/112880840-f97bc900-9098-11eb-9c2d-787f0eb9c642.png)
but it can be converted into a zip and then extracted into 

![image](https://user-images.githubusercontent.com/72760579/112880947-1b754b80-9099-11eb-978c-07e9a7f08b24.png)

**Comment by jeffalo** [comment]
this is a really good idea, this solves the issue of collaborative project editing and can be used incase something breaks and you want to rollback your project a couple versions.

this is, however only as good as the implementation. 

should it be actually github based, or should it be general purpose git based?

**Comment by FunctionalMetatable** [comment]
how would that work though:- would you guys need to use OAuth or something?

**Comment by RokCoder** [comment]
I was thinking about this some more and started dabbling with some backend code towards the goal. It's pretty much there (though I should note that 'm skipping over any legacy SB2 opcodes). I'm happy to neaten off the rough edges if someone fancies hooking up the Git integration side. www.rokcoder.com/sb3-to-txt

**Comment by RokCoder** [comment]
I'm not entirely sure that add-ons is ideally suited to adding github (or other) integration. I've set up source control locally for myself now with details being here - https://scratch.mit.edu/discuss/topic/509008. If someone feels that there's a place for this in add-ons then I'm open to discussion but this thread seemed to die off on its own so I'm assuming there is no real interest.

**Reply by TheColaber** [reply]
I think that this idea would be better as a whole different extension rather than an addon here, because of integration with github (or others).
